### PR TITLE
[DSSv3] Fix Issue 16210 - std.utf.byUTF can be made into a bidirectional range

### PR DESCRIPTION
**Pull request by vladchicos**
I also added unittests and an example.

### CODE REVIEW

File: std/utf.d [hunk 1]
```diff
@@ -4369,10 +4381,45 @@ if (isSomeChar!C)
                     }
                 }
 
+                static if (isBidirectionalRange!R)
+                {
+                    @property dchar back() scope // 'scope'' required by call to decodeBack() below
+                    {
+                        if (backBuff == Empty)
```
Comment (thewilsonator) [inline std/utf.d:4388]:
```
if (backBuff != Empty)
    return cast(dchar) backBuff;
...
return cast(dchar) backBuff;

File: std/utf.d [hunk 2]
```diff
@@ -4446,10 +4510,53 @@ if (isSomeChar!C)
                     }
                 }
 
+                static if (isBidirectionalRange!R)
+                {
+                    @property auto back() scope // 'scope' required by call to decodeBack() below
+                    {
+                        if (backPos == backFill)
```
Comment (thewilsonator) [inline std/utf.d:4517]: ditto

**Review by thewilsonator — COMMENTED** [review]
please use early returns

File: std/utf.d (hunk 1 above)
Comment (vladchicos) [reply std/utf.d]: I think I solved it.

**Review by RazvanN7 — APPROVED** [review]

File: std/utf.d [hunk 3]
```diff
@@ -4278,9 +4278,9 @@ private int impureVariable;
  *      Does not use GC if `useReplacementDchar` is set to `UseReplacementDchar.no`
  *
  * Returns:
- *      A forward range if `R` is a range and not auto-decodable, as defined by
+ *      A bidirectional range if `R` is a range and not auto-decodable, as defined by
  *      $(REF isAutodecodableString, std, traits), and if the base range is
- *      also a forward range.
+ *      also a bidirectional range.
```
Comment (RazvanN7) [inline std/utf.d:4283]: You are omitting the fact that if the underlying range is a forward range, a forward range is returned.

File: std/utf.d [hunk 4]
```diff
@@ -4325,10 +4325,22 @@ if (isSomeChar!C)
                     this.buff = buff;
                 }
 
+                static if (isBidirectionalRange!R)
+                {
+                    this(return R r, uint frontBuff, uint backBuff)
+                    {
+                        this.r = r;
+                        this.buff = frontBuff;
+                        this.backBuff = backBuff;
```
Comment (RazvanN7) [inline std/utf.d:4334]: Is this constructor needed? I don't see any calls to it. Also, it seems that it is not covered by tests.

File: std/utf.d [hunk 5]
```diff
@@ -4394,9 +4441,26 @@ if (isSomeChar!C)
                     this.buf = buf;
                 }
 
+                static if (isBidirectionalRange!R)
+                {
+                    this(return R r, ushort frontPos, ushort frontFill,
+                         ushort backPos, ushort backFill, C[4 / C.sizeof] buf)
```
Comment (RazvanN7) [inline std/utf.d:4447]: Again, I don't see any calls to this constructor. Is this needed?

**Review by RazvanN7 — CHANGES_REQUESTED** [review]

**Comment by RazvanN7** [comment]
ping @vladchicos

**Comment by RazvanN7** [comment]
This looks like a rebase screw up.

File: std/utf.d (hunk 5 above)
Comment (vladchicos) [reply std/utf.d:4455]: @RazvanN7  ,sorry for the delay and thanks for pointing it out. Right now the constructor is not used, but I discovered a problem in the meantime.
Consider this piece of code
```
wchar[] s = ['Ă', 'ț'];
auto r = s.byUTF!char;
r.popBack;
auto rCopy = r.save
//right now r == [0xc4, 0x82, 0xc8];
//and rCopy == [0xc4, 0x82];
```
It is fixable by having different save functions for forward and bidirectional ranges.
(And this way I will need the new constructor).
Is it ok if I do this, right?
I also pushed a commit to see what I mean, now I have to fix the rebasing.

File: std/utf.d (hunk 3 above)
Comment (vladchicos) [reply std/utf.d]: @RazvanN7 fixed it.

File: std/file.d [hunk 6]
```diff
@@ -402,24 +402,24 @@ version (Posix) private void[] readImpl(scope const(char)[] name, scope const(FS
...
                               SECURITY_ATTRIBUTES *lpSecurityAttributes, DWORD dwCreationDisposition,
-                              DWORD dwFlagsAndAttributes, HANDLE hTemplateFile) @trusted
+                              DWORD dwFlagsAndAttributes, HANDLE hTemplateFile)
     {
         return CreateFileW(namez, dwDesiredAccess, dwShareMode,
                            lpSecurityAttributes, dwCreationDisposition,
                            dwFlagsAndAttributes, hTemplateFile);
 
     }
-    static trustedCloseHandle(HANDLE hObject) @trusted
```
Comment (RazvanN7) [inline std/file.d:418]: This is not right.

File: std/utf.d (hunk 5 above)
Comment (RazvanN7) [reply std/utf.d:4455]: I see, so the saved back buff is not considered in the initial version of `save`.

File: std/utf.d [hunk 7]
```diff
@@ -4361,18 +4374,60 @@ if (isSomeChar!C)
                     buff = Empty;
                 }
 
-                static if (isForwardRange!R)
+                static if (isBidirectionalRange!R)
```
Comment (RazvanN7) [inline std/utf.d:4377]: you could move the `static if` inside the save function:

```d
static if (isForwardRange!R)
{
    auto save()
    {
         static if (isBidirectionalRange!R) { ... }
         else  { ... }
    }

Comment (vladchicos) [reply std/utf.d]: @RazvanN7 I changed it, I fixed the rebase and I added an unittest for the save() function. I'm sorry for the rebasing mistake.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**